Add BGP Tests: Phase1

tssurya opened the pull request:
## 📑 Description

This PR adds missing e2e test coverage for BGP.
Predominantly:

1. 2 shard conformance lanes i.e lgw and sgw to run regression tests on BGP exposed default network
2. 2 CP lanes i.e lgw,dual and sgw,v4 to run pod->external traffic flow test with the default network
3. importing routes is tested as part of the exporting routes already

In future we will build upon this, but let's take the first step here and then make incremental changes.

tssurya commented:
```
Summarizing 1 Failure:
  [FAIL] Node Operations on delete [It] deletes iptables rules with ExternalIP
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/gateway_localnet_linux_test.go:1511

Ran 187 of 187 Specs in 99.786 seconds
FAIL! -- 186 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestNodeSuite (99.96s)
=== RUN   TestOpenFlowManagerDefaultNetOVSBridgeFinder
=== RUN   TestOpenFlowManagerDefaultNetOVSBridgeFinder/empty_input_ports
=== RUN   TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_without_patch_ports
=== RUN   TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_with_a_patch_port
=== RUN   TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_with_a_patch_port_for_a_localnet_network
=== RUN   TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_with_a_patch_port_for_the_default_network_and_a_localnet
=== RUN   TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_with_a_patch_port_for_the_default_network,_a_localnet,_and_an_extra_primary_UDN
--- PASS: TestOpenFlowManagerDefaultNetOVSBridgeFinder (0.00s)
    --- PASS: TestOpenFlowManagerDefaultNetOVSBridgeFinder/empty_input_ports (0.00s)
    --- PASS: TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_without_patch_ports (0.00s)
    --- PASS: TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_with_a_patch_port (0.00s)
    --- PASS: TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_with_a_patch_port_for_a_localnet_network (0.00s)
    --- PASS: TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_with_a_patch_port_for_the_default_network_and_a_localnet (0.00s)
    --- PASS: TestOpenFlowManagerDefaultNetOVSBridgeFinder/input_ports_with_a_patch_port_for_the_default_network,_a_localnet,_and_an_extra_primary_UDN (0.00s)
FAIL
coverage: 53.3% of statements
make: *** [Makefile:66: check] Error 1
Error: Process completed with exit code 2.
```
UGH https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/13903582870/job/38901140783?pr=5121

tssurya commented on contrib/kind-common:
```diff
@@ -676,19 +676,86 @@ deploy_frr_external_container() {
+  # in the e2e tests that levergae router advertisements.
+  # This container will be connected to the frr container deployed above to simulate a realistic
+  # network topology
+  # -----------------               ------------------                         ---------------------
```
ugh fix formatting

tssurya replied:
done

qinqon commented on .github/workflows/test.yml:
```diff
@@ -514,6 +520,30 @@ jobs:
           echo OVN_TEST_EX_GW_NETWORK=kindexgw >> $GITHUB_ENV
           echo OVN_ENABLE_EX_GW_NETWORK_BRIDGE=true >> $GITHUB_ENV
         fi
+        if [[ "$JOB_NAME" == *"shard-conformance"* ]] && [ "$ADVERTISE_DEFAULT_NETWORK" == "true" ]; then
```
We cannot have JOB_NAME at kind.sh

tssurya replied:
yup yup we can just need to check if its set or not :) https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5121#discussion_r2000564580

tssurya replied:
this is done

tssurya commented on contrib/kind-common:
```diff
@@ -703,9 +770,57 @@ install_ffr_k8s() {
 
   rm -rf "${FRR_TMP_DIR}"
+  # Add routes for pod networks dynamically into the github runner for return traffic to pass back
+  if [[ "$JOB_NAME" == *"shard-conformance"* ]] && [ "$ADVERTISE_DEFAULT_NETWORK" == "true" ]; then
```
done

tssurya commented:
YAYYYY FINALLY GETTING GREEN CI
side-note: i don't know why I ran into issues accessing kube-api server from github runner in the workflow step but ultimately i added that logic into the kind script itself

tssurya commented:
/retest-failed

martinkennelly reviewed (commented):
looking good

tssurya commented on .github/workflows/test.yml:
```diff
@@ -592,6 +616,22 @@ jobs:
           make -C test traffic-flow-tests WHAT="run"
         fi
 
+    - name: Check iptables and routes
```
yes! two reasons:

sudo ip r shows the routes I am adding back for the BGP lane are indeed present and the rules show the MASQUERADE count going up.. it might be a trivial step today but in the future useful for debugging if something changes which hopefully it won't

tssurya replied:
in the end I removed this

tssurya commented on test/scripts/e2e-cp.sh:
```diff
@@ -160,6 +160,15 @@ if [[ "${WHAT}" != "${NETWORK_SEGMENTATION_TESTS}"* ]]; then
   SKIPPED_TESTS+=$NETWORK_SEGMENTATION_TESTS
 fi
 
+# Only run network segmentation tests if they are explicitly requested
```
hehe yes copy-paste snafu..nice catch

tssurya replied:
fixed it

tssurya commented:
https://github.com/metallb/frr-k8s/issues/261 should fix the flake: https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/13922536091/job/38960790242?pr=5121

tssurya commented:
@martinkennelly I addressed your comments in this push: https://github.com/ovn-kubernetes/ovn-kubernetes/compare/c62a96aaab7e07f6d0ce8909fcc866d2ba6bd6e1..b7bc838177baeb25378ba6104dad8cc4b6fb2ded

and then I pushed again to resolve a minor issue with the patches we were consuming which is required for the next UDN test PR to work properly: https://github.com/ovn-kubernetes/ovn-kubernetes/compare/b7bc838177baeb25378ba6104dad8cc4b6fb2ded..cbab4178cb21885ff1fbae8426e8d0b9a59d4318

no new change in functional logic.. I think we are almost good here to merge this PR PTAL.

tssurya commented:
@martinkennelly CI is golden
the network-segmentation lanes are failing due to:
```
Summarizing 1 Failure:
  [FAIL] e2e egress IP validation Network Segmentation: IPv4 L3 role primary [OVN network] Using different methods to disable a node's availability for egress Should validate the egress IP functionality against remote hosts [It] disabling egress nodes impeding GRCP health check
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/egressip.go:479
```
known flakes unrelated to this PR
